File: mmd_generator/exporter.py
"""Export .mmd, tag list and review report (Sections 34-36)."""
from __future__ import annotations

import csv
import io
from pathlib import Path
from typing import Union

from .models import GenerationResult, ReadinessGate, Severity

PathLike = Union[str, Path]


def export_mmd(path: PathLike, result: GenerationResult) -> Path:
    p = Path(path)
    p.write_text(result.mmd_text, encoding="utf-8")
    return p


def export_tag_list(path: PathLike, result: GenerationResult) -> Path:
    """Section 35: Tag Name / MMD Syntax / Type / Occurrences / Source Text
    / Status / Notes, as CSV."""
    p = Path(path)
    with p.open("w", newline="", encoding="utf-8") as fh:
        writer = csv.writer(fh)
        writer.writerow(
            ["Tag Name", "MMD Syntax", "Type", "Occurrences", "Source Text", "Status", "Notes"]
        )
        for tag in result.tags:
            source = tag.source_texts[0] if tag.source_texts else ""
            writer.writerow(
                [tag.name, tag.mmd_syntax, tag.tag_type.value, tag.occurrences, source, tag.status.value, tag.notes]
            )
        writer.writerow([])
        writer.writerow(["Conditional", "MMD Open", "MMD Close", "Occurrences", "Scope", "Confirmed", "Status"])
        for cond in result.conditionals:
            writer.writerow(
                [
                    cond.name,
                    cond.mmd_open,
                    cond.mmd_close,
                    cond.occurrences,
                    cond.scope,
                    "Yes" if cond.confirmed else "No",
                    cond.status,
                ]
            )
    return p


_GATE_ICON = {
    ReadinessGate.NOT_READY: "\U0001F534 NOT READY",
    ReadinessGate.REVIEW_REQUIRED: "\U0001F7E0 REVIEW REQUIRED",
    ReadinessGate.READY: "\U0001F7E2 READY",
}


def build_review_report(result: GenerationResult) -> str:
    buf = io.StringIO()
    buf.write("# LETTER VALIDATION REPORT\n\n")
    buf.write(f"Letter type: {result.letter_type.value}\n")
    if result.detected_type:
        buf.write(f"Auto-detected type: {result.detected_type.value}\n")
    buf.write(f"Export status: {_GATE_ICON[result.readiness]}\n")
    buf.write(f"Page layout confidence: {result.layout_confidence:.0f}%\n\n")

    errors = [i for i in result.validation_issues if i.severity == Severity.ERROR]
    warnings = [i for i in result.validation_issues if i.severity == Severity.WARNING]

    buf.write("## Errors\n")
    if errors:
        for i in errors:
            buf.write(f"- [{i.code}] {i.message}\n")
    else:
        buf.write("(none)\n")

    buf.write("\n## Warnings\n")
    if warnings:
        for i in warnings:
            buf.write(f"- [{i.code}] {i.message}\n")
    else:
        buf.write("(none)\n")

    buf.write("\n## Page Layout Warnings\n")
    for w in result.layout_warnings:
        prefix = "OK" if w.severity == Severity.INFO else "WARN"
        buf.write(f"- [{prefix}] {w.message}\n")

    new_tags = [t for t in result.tags if t.status.value == "Needs Creation"]
    dup_tags = [t for t in result.tags if t.status.value == "Duplicate"]
    review_tags = [t for t in result.tags if t.status.value == "Review"]

    buf.write("\n## New Data Tags\n")
    for t in new_tags:
        buf.write(f"- {t.mmd_syntax} (x{t.occurrences})\n")
    if not new_tags:
        buf.write("(none)\n")

    buf.write("\n## Potential Duplicate Tags\n")
    for t in dup_tags:
        buf.write(f"- {t.name}: {t.notes}\n")
    if not dup_tags:
        buf.write("(none)\n")

    buf.write("\n## Tags Requiring Review\n")
    for t in review_tags:
        buf.write(f"- {t.mmd_syntax or '(unresolved)'} : {t.notes or 'context could not be determined'}\n")
    if not review_tags:
        buf.write("(none)\n")

    buf.write("\n## Unresolved Conditions\n")
    unresolved = [c for c in result.conditionals if not c.confirmed]
    for c in unresolved:
        buf.write(f"- {c.name} (scope: {c.scope}, occurrences: {c.occurrences})\n")
    if not unresolved:
        buf.write("(none)\n")

    return buf.getvalue()


def export_review_report(path: PathLike, result: GenerationResult) -> Path:
    p = Path(path)
    p.write_text(build_review_report(result), encoding="utf-8")
    return p


def export_all(outdir: PathLike, result: GenerationResult, base_name: str = "letter") -> dict:
    outdir = Path(outdir)
    outdir.mkdir(parents=True, exist_ok=True)
    return {
        "mmd": export_mmd(outdir / f"{base_name}.mmd", result),
        "tags": export_tag_list(outdir / f"{base_name}_tags.csv", result),
        "report": export_review_report(outdir / f"{base_name}_review_report.md", result),
    }

File: mmd-dashboard/lib/exporter.ts
// Export .mmd, tag list (CSV) and review report (Markdown).
// Port of mmd_generator/exporter.py -- returns strings for the browser to
// download rather than writing files to disk.

import type { GenerationResult, ReadinessGate } from "./models";

function csvEscape(value: string | number): string {
  const s = String(value);
  if (/[",\n]/.test(s)) return `"${s.replace(/"/g, '""')}"`;
  return s;
}

export function buildTagListCsv(result: GenerationResult): string {
  const rows: string[] = [];
  rows.push(["Tag Name", "MMD Syntax", "Type", "Occurrences", "Source Text", "Status", "Notes"].map(csvEscape).join(","));
  for (const tag of result.tags) {
    const source = tag.sourceTexts[0] ?? "";
    rows.push(
      [tag.name, tag.mmdSyntax, tag.tagType, tag.occurrences, source, tag.status, tag.notes]
        .map(csvEscape)
        .join(","),
    );
  }
  rows.push("");
  rows.push(["Conditional", "MMD Open", "MMD Close", "Occurrences", "Scope", "Confirmed", "Status"].map(csvEscape).join(","));
  for (const cond of result.conditionals) {
    rows.push(
      [cond.name, cond.mmdOpen, cond.mmdClose, cond.occurrences, cond.scope, cond.confirmed ? "Yes" : "No", cond.status]
        .map(csvEscape)
        .join(","),
    );
  }
  return rows.join("\n");
}

const GATE_ICON: Record<ReadinessGate, string> = {
  "NOT READY": "\u{1F534} NOT READY",
  "REVIEW REQUIRED": "\u{1F7E0} REVIEW REQUIRED",
  READY: "\u{1F7E2} READY",
};

export function buildReviewReport(result: GenerationResult): string {
  const lines: string[] = [];
  lines.push("# LETTER VALIDATION REPORT", "");
  lines.push(`Letter type: ${result.letterType}`);
  if (result.detectedType) lines.push(`Auto-detected type: ${result.detectedType}`);
  lines.push(`Export status: ${GATE_ICON[result.readiness]}`);
  lines.push(`Page layout confidence: ${result.layoutConfidence.toFixed(0)}%`, "");

  const errors = result.validationIssues.filter((i) => i.severity === "ERROR");
  const warnings = result.validationIssues.filter((i) => i.severity === "WARNING");

  lines.push("## Errors");
  lines.push(...(errors.length ? errors.map((i) => `- [${i.code}] ${i.message}`) : ["(none)"]));

  lines.push("", "## Warnings");
  lines.push(...(warnings.length ? warnings.map((i) => `- [${i.code}] ${i.message}`) : ["(none)"]));

  lines.push("", "## Page Layout Warnings");
  lines.push(...result.layoutWarnings.map((w) => `- [${w.severity === "INFO" ? "OK" : "WARN"}] ${w.message}`));

  const newTags = result.tags.filter((t) => t.status === "Needs Creation");
  const dupTags = result.tags.filter((t) => t.status === "Duplicate");
  const reviewTags = result.tags.filter((t) => t.status === "Review");

  lines.push("", "## New Data Tags");
  lines.push(...(newTags.length ? newTags.map((t) => `- ${t.mmdSyntax} (x${t.occurrences})`) : ["(none)"]));

  lines.push("", "## Potential Duplicate Tags");
  lines.push(...(dupTags.length ? dupTags.map((t) => `- ${t.name}: ${t.notes}`) : ["(none)"]));

  lines.push("", "## Tags Requiring Review");
  lines.push(
    ...(reviewTags.length
      ? reviewTags.map((t) => `- ${t.mmdSyntax || "(unresolved)"} : ${t.notes || "context could not be determined"}`)
      : ["(none)"]),
  );

  const unresolved = result.conditionals.filter((c) => !c.confirmed);
  lines.push("", "## Unresolved Conditions");
  lines.push(
    ...(unresolved.length
      ? unresolved.map((c) => `- ${c.name} (scope: ${c.scope}, occurrences: ${c.occurrences})`)
      : ["(none)"]),
  );

  return lines.join("\n") + "\n";
}

export function downloadTextFile(filename: string, content: string, mime = "text/plain"): void {
  const blob = new Blob([content], { type: mime });
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = filename;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
}
